# Day 8: Seven Segment Search

## Part one

You barely reach the safety of the cave when the whale smashes into the cave mouth, collapsing it. Sensors indicate another exit to this cave at a much greater depth, so you have no choice but to press on.

As your submarine slowly makes its way through the cave system, you notice that the four-digit *seven-segment displays* in your submarine are malfunctioning; they must have been damaged during the escape. You'll be in a lot of trouble without them, so you'd better figure out what's wrong.

Each digit of a seven-segment display is rendered by turning on or off any of seven segments named a through g:

```
  0:      1:      2:      3:      4:
 aaaa    ....    aaaa    aaaa    ....
b    c  .    c  .    c  .    c  b    c
b    c  .    c  .    c  .    c  b    c
 ....    ....    dddd    dddd    dddd
e    f  .    f  e    .  .    f  .    f
e    f  .    f  e    .  .    f  .    f
 gggg    ....    gggg    gggg    ....

  5:      6:      7:      8:      9:
 aaaa    aaaa    aaaa    aaaa    aaaa
b    .  b    .  .    c  b    c  b    c
b    .  b    .  .    c  b    c  b    c
 dddd    dddd    ....    dddd    dddd
.    f  e    f  .    f  e    f  .    f
.    f  e    f  .    f  e    f  .    f
 gggg    gggg    ....    gggg    gggg
```

So, to render a `1`, only segments `c` and `f` would be turned on; the rest would be off. To render a `7`, only segments `a`, `c`, and `f` would be turned on.

The problem is that the signals which control the segments have been mixed up on each display. The submarine is still trying to display numbers by producing output on signal wires `a` through `g`, but those wires are connected to segments **randomly**. Worse, the wire/segment connections are mixed up separately for each four-digit display! (All of the digits **within** a display use the same connections, though.)

So, you might know that only signal wires b and g are turned on, but that doesn't mean **segments** `b` and `g` are turned on: the only digit that uses two segments is `1`, so it must mean segments c and f are meant to be on. With just that information, you still can't tell which wire (`b`/`g`) goes to which segment (`c`/`f`). For that, you'll need to collect more information.

For each display, you watch the changing signals for a while, make a note of **all ten unique signal patterns** you see, and then write down a single **four digit output value** (your puzzle input). Using the signal patterns, you should be able to work out which pattern corresponds to which digit.

For example, here is what you might see in a single entry in your notes:

```
acedgfb cdfbe gcdfa fbcad dab cefabd cdfgeb eafb cagedb ab |
cdfeb fcadb cdfeb cdbaf
```

(The entry is wrapped here to two lines so it fits; in your notes, it will all be on a single line.)

Each entry consists of ten **unique signal patterns**, a | delimiter, and finally the **four digit output value**. Within an entry, the same wire/segment connections are used (but you don't know what the connections actually are). The unique signal patterns correspond to the ten different ways the submarine tries to render a digit using the current wire/segment connections. Because `7` is the only digit that uses three segments, dab in the above example means that to render a `7`, signal lines d, a, and b are on. Because `4` is the only digit that uses four segments, eafb means that to render a `4`, signal lines e, a, f, and b are on.

Using this information, you should be able to work out which combination of signal wires corresponds to each of the ten digits. Then, you can decode the four digit output value. Unfortunately, in the above example, all of the digits in the output value `(cdfeb fcadb cdfeb cdbaf)` use five segments and are more difficult to deduce.

For now, **focus on the easy digits**. Consider this larger example:

```
be cfbegad cbdgef fgaecd cgeb fdcge agebfd fecdb fabcd edb |
fdgacbe cefdb cefbgd gcbe
edbfga begcd cbg gc gcadebf fbgde acbgfd abcde gfcbed gfec |
fcgedb cgb dgebacf gc
fgaebd cg bdaec gdafb agbcfd gdcbef bgcad gfac gcb cdgabef |
cg cg fdcagb cbg
fbegcd cbd adcefb dageb afcb bc aefdc ecdab fgdeca fcdbega |
efabcd cedba gadfec cb
aecbfdg fbg gf bafeg dbefa fcge gcbea fcaegb dgceab fcbdga |
gecf egdcabf bgf bfgea
fgeab ca afcebg bdacfeg cfaedg gcfdb baec bfadeg bafgc acf |
gebdcfa ecba ca fadegcb
dbcfg fgd bdegcaf fgec aegbdf ecdfab fbedc dacgb gdcebf gf |
cefg dcbef fcge gbcadfe
bdfegc cbegaf gecbf dfcage bdacg ed bedf ced adcbefg gebcd |
ed bcgafe cdgba cbgef
egadfb cdbfeg cegd fecab cgb gbdefca cg fgcdab egfdb bfceg |
gbdfcae bgc cg cgb
gcafb gcf dcaebfg ecagb gf abcdeg gaef cafbge fdbac fegbdc |
fgae cfgab fg bagce
```

Because the digits `1`, `4`, `7`, and `8` each use a unique number of segments, you should be able to tell which combinations of signals correspond to those digits. **Counting only digits in the output values** (the part after | on each line), in the above example, there are **`26`** instances of digits that use a unique number of segments (highlighted above).

**In the output values, how many times do digits `1`, `4`, `7`, or `8` appear?**

In [1]:
ONE = 2
FOUR = 4
SEVEN = 3
EIGHT = 7

outputs = [0, 0, 0, 0, 0, 0, 0, 0, 0]

with open('data/input.txt', 'r') as f:
    for line in f.read().splitlines():
        input_str, output_str = line.split(' | ')
        output_digits = output_str.split(' ')
        for output_digit in output_digits:
            output_digit = output_digit.strip()
            outputs[len(output_digit)] += 1
            
outputs

[0, 0, 78, 72, 87, 266, 231, 66, 0]

In [2]:
outputs[ONE] + outputs[FOUR] + outputs[SEVEN] + outputs[EIGHT]

303

Your puzzle answer was `303`.

## Part two

Through a little deduction, you should now be able to determine the remaining digits. Consider again the first example above:

```
acedgfb cdfbe gcdfa fbcad dab cefabd cdfgeb eafb cagedb ab |
cdfeb fcadb cdfeb cdbaf
```

After some careful analysis, the mapping between signal wires and segments only make sense in the following configuration:

```
 dddd
e    a
e    a
 ffff
g    b
g    b
 cccc
```

So, the unique signal patterns would correspond to the following digits:

- acedgfb: `8`
- cdfbe: `5`
- gcdfa: `2`
- fbcad: `3`
- dab: `7`
- cefabd: `9`
- cdfgeb: `6`
- eafb: `4`
- cagedb: `0`
- ab: `1`

Then, the four digits of the output value can be decoded:

- cdfeb: **`5`**
- fcadb: **`3`**
- cdfeb: **`5`**
- cdbaf: **`3`**

Therefore, the output value for this entry is **`5353`**.

Following this same process for each entry in the second, larger example above, the output value of each entry can be determined:

- fdgacbe cefdb cefbgd gcbe: `8394`
- fcgedb cgb dgebacf gc: `9781`
- cg cg fdcagb cbg: `1197`
- efabcd cedba gadfec cb: `9361`
- gecf egdcabf bgf bfgea: `4873`
- gebdcfa ecba ca fadegcb: `8418`
- cefg dcbef fcge gbcadfe: `4548`
- ed bcgafe cdgba cbgef: `1625`
- gbdfcae bgc cg cgb: `8717`
- fgae cfgab fg bagce: `4315`

Adding all of the output values in this larger example produces `61229`.

For each entry, determine all of the wire/segment connections and decode the four-digit output values. **What do you get if you add up all of the output values?**

In [3]:
ONE = 2
FOUR = 4
SEVEN = 3
EIGHT = 7

outputs = [0, 0, 0, 0, 0, 0, 0, 0, 0]

with open('data/input.txt', 'r') as f:
    for line in f.read().splitlines():
        input_str, output_str = line.split(' | ')
        output_digits = output_str.split(' ')
        for output_digit in output_digits:
            output_digit = output_digit.strip()
            outputs[len(output_digit)] += 1
            
outputs

tab = ['d', 'a', 'b']
tab.sort()
tab

['a', 'b', 'd']

In [4]:
class Digit:
    def __init__(self, char_seq):
        splitted_input_value = [char for char in char_seq]
        self.char_seq_set = set(splitted_input_value)
        self.length = len(self.char_seq_set)
        
        self.number = None
        
        self._guess_simple_number()
        
    def _guess_simple_number(self):
        if self.length == 2:
            self.number = 1
        elif self.length == 4:
            self.number = 4
        elif self.length == 3:
            self.number = 7
        elif self.length == 7:
            self.number = 8
            
    def match(self, digit):
        return self.char_seq_set == digit.char_seq_set
    
    def str_number(self):
        return str(self.number)

In [5]:
def prepare_input_mapping(input_line):
    digit_map = {
        'classified_digits': [
            None,
            None,
            None,
            None,
            None,
            None,
            None,
            None,
            None,
            None
        ],
        'unclassified_digits': []
    }

    for input in input_line.split():
        digit = Digit(input)
        if digit.number is None:
            digit_map['unclassified_digits'].append(digit)
        else:
            digit_map['classified_digits'][digit.number] = digit

    return digit_map

def map_input_digits(digit_map):
    wires = {
        'a': None,
        'b': None,
        'c': None,
        'd': None,
        'e': None,
        'f': None,
        'g': None
    }

    # Find 9
    wires['a'] = (digit_map['classified_digits'][7].char_seq_set - 
                  digit_map['classified_digits'][1].char_seq_set).pop()
    seven_four_set = digit_map['classified_digits'][7].char_seq_set | digit_map['classified_digits'][4].char_seq_set
    for digit in digit_map['unclassified_digits']:
        if digit.length == 6 and len(digit.char_seq_set - seven_four_set) == 1:
            digit.number = 9
            break
    digit_map['unclassified_digits'].remove(digit)
    digit_map['classified_digits'][digit.number] = digit
    wires['e'] = (digit_map['classified_digits'][8].char_seq_set - 
                  digit_map['classified_digits'][9].char_seq_set).pop()
    wires['g'] = (digit_map['classified_digits'][9].char_seq_set - seven_four_set).pop()

    # Find 2
    for digit in digit_map['unclassified_digits']:
        if digit.length == 5 and {wires['a'], wires['e'], wires['g']}.issubset(digit.char_seq_set):
            digit.number = 2
            break
    digit_map['unclassified_digits'].remove(digit)
    digit_map['classified_digits'][digit.number] = digit
    wires['d'] = (digit_map['classified_digits'][2].char_seq_set - 
                  (digit_map['classified_digits'][7].char_seq_set | { wires['g'], wires['e']})).pop()

    # Find 0 and 6
    for digit in digit_map['unclassified_digits']:
        if digit.length == 6 and wires['d'] not in digit.char_seq_set:
            digit.number = 0
            break
    digit_map['unclassified_digits'].remove(digit)
    digit_map['classified_digits'][digit.number] = digit

    # Find 6
    for digit in digit_map['unclassified_digits']:
        if digit.length == 6:
            digit.number = 6
            break
    digit_map['unclassified_digits'].remove(digit)
    digit_map['classified_digits'][digit.number] = digit
    wires['c'] = (digit_map['classified_digits'][8].char_seq_set - digit_map['classified_digits'][6].char_seq_set).pop()

    # Find 3 and 5
    for digit in digit_map['unclassified_digits']:
        if digit.length == 5 and wires['c'] in digit.char_seq_set:
            digit.number = 3
            break
    digit_map['unclassified_digits'].remove(digit)
    digit_map['classified_digits'][digit.number] = digit
    digit = digit_map['unclassified_digits'].pop()
    digit.number = 5
    digit_map['classified_digits'][digit.number] = digit

    return digit_map

In [6]:
digit_map = prepare_input_mapping('acedgfb cdfbe gcdfa fbcad dab cefabd cdfgeb eafb cagedb ab')
mapped_input_digits = map_input_digits(digit_map)
mapped_input_digits

{'classified_digits': [<__main__.Digit at 0x106f2d160>,
 'unclassified_digits': []}

In [7]:
def output_mapping(output_line, mapped_input_digits):
    output_digits = []
    for output in output_line.split():
        digit = Digit(output)
        if digit.number is None:
            for input_digit in mapped_input_digits['classified_digits']:
                if digit.match(input_digit):
                    digit.number = input_digit.number
                    break
        output_digits.append(digit)
    
    return output_digits

In [8]:
output_digits = output_mapping('cdfeb fcadb cdfeb cdbaf', mapped_input_digits)

res = ''
for output_digit in output_digits:
    res += output_digit.str_number()
int(res)

5353

In [9]:
res = 0

with open('data/input.txt', 'r') as f:
    for line in f.read().splitlines():
        input_line, output_line = line.split(' | ')
        
        digit_map = prepare_input_mapping(input_line)
        mapped_input_digits = map_input_digits(digit_map)
        
        output_digits = output_mapping(output_line, mapped_input_digits)
        
        str_digits = ''
        for output_digit in output_digits:
            str_digits += output_digit.str_number()
            
        res += int(str_digits)
        
res

961734

Your puzzle answer was `961734`.